examples/MQ/pixelDetector/src/devices/FairMQSimDevice.h

```diff
@@ -69,7 +69,7 @@ class FairMQSimDevice : public FairMQRunDevice
 
     bool fRunInitialized;   // false, set to true after initialization in the run stage (!)
 
-    FairRunSim* fRunSim;
```

Comment:
Why not initialise the member already here?

std::unique_ptr<FairRunSim> fRunSim{new FairRunSim()};

@dennisklein,

or must it be created in the InitTask method?

Reply:
We could move even to a `FairRunSim fRunSim;` and use value semantics. I considered that.

But…

I am planning to create a `FairExampleRunSim` that derives from `FairRunSim` to show that deriving is fully okay and that it might be useful to implement standard beheviour for an experiment. One of my ideas for this: The constructor takes an `mcEngine` parameter to set the transport engine. the transport engine isn't known in the ctor of `FairMQSimDevice`. So constructing the `FairExampleRunSim` wouldn't work.

(p.s. I actually have that all already mostly prepared. It's just waiting in a local branch for me to take another look at it.)

Reply:
> or must it be created in the InitTask method?

I wouldn't say "must", but yes, one should consider the lifetimes. In the [FairMQ state machine](https://github.com/FairRootGroup/FairMQ/blob/master/docs/Device.md#13-state-machine), the application should be configured and initialized in `INITIALIZING TASK` and cleaned in `RESETTING TASK`. As `FairRunSim` seems to be a default constructed placeholder object which gets its configuration supplied after construction anyways, I guess, it does not really matter *when* you create it. The only reason to not construct it with the Device would be, if it had side effects that would better only happen during `INITIALIZING TASK`.

Reply:
> I am planning to create a `FairExampleRunSim` that derives from `FairRunSim` to show that deriving is fully okay and that it might be useful to implement standard beheviour for an experiment. One of my ideas for this: The constructor takes an `mcEngine` parameter to set the transport engine. the transport engine isn't known in the ctor of `FairMQSimDevice`. So constructing the `FairExampleRunSim` wouldn't work.

:+1:

Reply:
> \[…\] In the [FairMQ state machine](https://github.com/FairRootGroup/FairMQ/blob/master/docs/Device.md#13-state-machine), the application should best be configured and initialized in `INITIALIZING TASK` and cleaned in `RESETTING TASK`. \[…\]

Following this, it sounds like we should `fRunSim.reset()` in `ResetTask` to properly clean up and get back to the state before InitTask.

Technically it wouldn't be needed, as `unique_ptr` takes care of most details for us. But it might be a "good citizen" thing. For example if there are two devices in the same process this would be needed (as there can only be one FairRun at any given time).

Should I add this?

Reply:
> Following this, it sounds like we should fRunSim.reset() in ResetTask to properly clean up and get back to the state before InitTask.

Sure, if it makes sense for the `FairRunSim`.

Alternatively, you could say a reset cycle is not supported and add some kind of an error, if devices go through a InitTask twice. (At the moment, it is not possible for a user to change the FMQ state machine.)

Reply:
IMHO FairRunSim should support all of this. So let's have a proper ResetTask that cleans up/etc.

I changed the commit accordingly.